uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-balance-report.component.js

```diff
@@ -1,17 +1,114 @@
 /* eslint-env amd */
+/* global angular */
 
 define([
-  'leave-absences/shared/modules/components'
-], function (components) {
+  'common/lodash',
+  'leave-absences/shared/modules/components',
+  'leave-absences/shared/models/leave-balance-report.model',
+  'common/services/notification'
+], function (_, components) {
   components.component('leaveBalanceReport', {
-    controller: balanceReportController,
+    controller: controller,
     controllerAs: 'vm',
     templateUrl: ['shared-settings', function (sharedSettings) {
       return sharedSettings.sharedPathTpl + 'components/leave-balance-report.html';
     }]
   });
 
-  function balanceReportController () {
+  controller.$inject = ['$q', 'AbsenceType', 'LeaveBalanceReport', 'notification', 'Session'];
 
+  function controller ($q, AbsenceType, LeaveBalanceReport, notification, Session) {
+    var loggedInContactId;
+    var pageSize = 50;
+    var vm = this;
+
+    vm.absenceTypes = [];
+    vm.loading = { report: false };
+    vm.report = [];
+    vm.reportCount = 0;
+
+    /**
+     * Initializes the component. Loads logged in contact information and
+     * the first page of the report.
+     */
+    vm.$onInit = function () {
+      vm.loading.report = true;
+
+      $q.all([
+        initAbsenceTypes(),
+        initLoggedInContactId()
+      ])
+      .then(function () {
+        return vm.loadReportPage(1);
+      })
+      .catch(function (error) {
+        notification.error('Error', error);
+      });
+    };
+
+    /**
+     * Loads a specific page of the report.
+     *
+     * @return {Promise}
+     */
+    vm.loadReportPage = function (pageNumber) {
+      pageNumber = pageNumber || 1;
+      vm.loading.report = true;
+
+      return LeaveBalanceReport.all(
+        { managed_by: loggedInContactId },
+        { page: pageNumber, size: pageSize }
+      ).then(function (response) {
+        vm.report = sortBlanceReportAbsenceTypes(response.list);
+        vm.reportCount = response.total;
+      })
+      .catch(function (error) {
+        notification.error('Error', error.error_message);
+      })
+      .finally(function () {
+        vm.loading.report = false;
+      });
+    };
+
+    /**
+     *
+     */
+    function initAbsenceTypes () {
+      return AbsenceType.all({options: { sort: 'title ASC' }})
+      .then(function (response) {
+        vm.absenceTypes = response;
+      });
+    }
+
+    /**
+     * Initializes the loggedInContactId using the session value.
+     *
+     * @return {Promise}
+     */
+    function initLoggedInContactId () {
+      return Session.get().then(function (value) {
+        loggedInContactId = value.contact_id;
+      });
+    }
+
+    /**
+     * Sorts the absence types of each balance report according to the order of
```

Comment:
This is very confusing yes, I was trying to explain with words but I think adding the final object structure would be more clear.

The problem here is the absence types column groups should be in alphabetical order:
![headers](https://user-images.githubusercontent.com/1642119/29271481-1d5d0c9e-80a8-11e7-8817-b39c57f5fa37.png)

For each record in the report, I need to display the entitlement, used, balance, and request values of the absence types, but this record may not be sorted the same as in the headers. The records have this structure:

```js
{
  contact_id: 1,
  contact_display_name: 'Jon Doe',
  absence_types: [
    { id: 1, values..... },
    { id: 2, values..... },
    { id: 3, values..... }
  ]
}
```

The order of the absence_types may be 1, 2, and 3, but I need them to be 2, 3, 1 (for example) so they match the order of the column group.